stan/math/prim/functor/multi_expression.hpp

```diff
@@ -0,0 +1,196 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_MULTI_EXPRESSION_HPP
+#define STAN_MATH_PRIM_FUNCTOR_MULTI_EXPRESSION_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err.hpp>
+#include <tuple>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+constexpr int static_sum() { return 0; }
+
+template <typename Arg0, typename... Args>
+constexpr int static_sum(Arg0 arg0, Args... args) {
+  return arg0 + static_sum(args...);
+}
+
+}  // namespace internal
+
+/**
+ * Represents multiple exprs that will be calculated in same loop.
+ * @tparam T_expressions types of exprs
+ */
+template <typename... T_expressions>
+class eigen_expressions_ {
+ public:
+  /**
+   * Constructor.
+   * @param exprs expresions that will be calculated in same loop.
+   * @throw invalid_argument expressions have different sizes (for row major
+   * expressions rows and columns are swapped for the check)
```

Comment:
Why does this size swapping make sense?

Reply:
Because in loops we iterate over inner and outer dimension - so here we are not comparing if the expressions have the same number of rows and cols, but actually if they have the same inner and outer dimension. Does that make sense?